### PR TITLE
[5.1] Added preserveKeys to Collection groupBy method

### PR DESCRIPTION
**Pull request by btpoe**
I made it an opt-in option, so it won't change the default behavior.

This is helpful when using `->keyBy()` prior to calling `->groupBy()`. The current behavior, calling the following will wipe out the previously set keys.

``` php
$result = $collection
  ->keyBy('id')
  ->groupBy('parent_id');

// yields
$result = [
  '2' => [
    '0' => [
      'id' => '1',
      'name' => 'foo',
      'parent_id' => '2'
    ]
  ],
  '3' => [
    '0' => [
      'id' => '2',
      'name' => 'bar',
      'parent_id' => '3'
    ]
  ]
]
```

Setting $preserveKeys to `true` will keep them in tact.

``` php
$result = $collection
  ->keyBy('id')
  ->groupBy('parent_id', true);

// yields
$result = [
  '2' => [
    '1' => [
      'id' => '1',
      'name' => 'foo',
      'parent_id' => '2'
    ]
  ],
  '3' => [
    '2' => [
      'id' => '2',
      'name' => 'bar',
      'parent_id' => '3'
    ]
  ]
]
```

### CODE REVIEW

File: src/Illuminate/Support/Collection.php [hunk 1]
```diff
@@ -252,7 +253,14 @@ public function groupBy($groupBy)
...
 
-			$results[$groupKey]->push($value);
+			if ($preserveKeys)
+			{
+				$results[$groupKey]->put($key, $value);
+			}
+			else
+			{
+				$results[$groupKey]->push($value);
+			}
```
Comment (JosephSilber) [inline src/Illuminate/Support/Collection.php]: You can shorten this to:

```php
$results[$groupKey]->offsetSet($preserveKeys ? $key : null, $value);
```

Comment (GrahamCampbell) [inline src/Illuminate/Support/Collection.php]: The one liner looks fine. :)

**Comment by btpoe** [comment]
@JosephSilber, I've applied your suggested change. Thanks!

File: src/Illuminate/Support/Collection.php [hunk 2]
```diff
@@ -252,7 +253,7 @@ public function groupBy($groupBy)
 				$results[$groupKey] = new static;
 			}
 
-			$results[$groupKey]->push($value);
+			$results[$groupKey]->offsetSet($preserveKeys ? $key : null, $value);
```
Comment (GrahamCampbell) [inline src/Illuminate/Support/Collection.php]: no trailing whitespace please

Comment (GrahamCampbell) [inline src/Illuminate/Support/Collection.php]: please squash to one commit too

**Comment by btpoe** [comment]
trimmed and squashed.

**Comment by taylorotwell** [comment]
Is there a big problem with just making this the default behavior?

**Comment by btpoe** [comment]
It could be the default, but someone might rely on zero-based indexes.
